### ISSUE TITLE
plugin/forward: queries are getting stuck in forward plugin during restart (reload)

### ISSUE THREAD

**Issue by rdrozhdzh** [coredns/coredns#1666]
During graceful restart (reload) the coredns calls registered `onShutdown()` handlers for the plugin objects to be terminated. Unfortunately, at this time server is not stopped yes and it continues accepting and processing new queries. So, for the same plugin `ServeDNS()` and `onShutdown()` can be called in any order.

Calling `onShutdown()` in forward plugin leads to stopping transport object, i.e. it stops `connManager()` method which reads `Dial` requests from channel. However, `Dial` request may still arrive after stopping transport, and such request hangs during write to channel, because nobody reads the channel on another side.

The problem looks even more serious for requests received via grpc connection. Because grpc server is stopped with `GracefulStop()` method (from `google.golang.org/grpc` package) which waits until all incoming queries get processed. Therefore, in this case `caddy.Restart()` function is not finished at all

The problem can be fixed on forward plugin side by dropping requests which come after call of `onShutdown()`. But I guess other plugins may have similar issues. Maybe it's worth to revise caddy's restart process and first shutdown servers (I mean stop receiving new requests and wait until all in-progress requests leave plugin's pipeline) and then call the registered `onShutdown()` handlers

**Comment by fturib** [comment]
Right now, the event sequence in Caddy is the event BEFORE the action:

START : onFirstStartup(), on Startup() .. then start the servers
RESTART: onRestart(),  onStartup(), start the new servers, onShutdown(), stop the old servers
STOP: (on SIGTERM/SIGQUIT) onShutdown(), onFinalShutdown(), then stop all servers

Looks like all the events are BEFORE the action on the servers.
Useful when you want to switch on the resources before the first Query. 
But for shutdown, if you switch down your resources you may not be able to serve up to the stop() of the servers.

WARNING: About reload, there is another drawback : if the start and Stop does not succeed, Caddy just stop to process the Restart. Depending WHERE is the error, you may end up with none of the new Server neither old server running correctly. (for instance if one of the Shutdown() is returning an error).

**Comment by johnbelamaric** [comment]
Don't the sockets get passed on to the new servers before onShutdown is called? If so, then ServeDNS should be in the new server not the old server. We may still want to have onShutdown wait for the pending requests to clear out. But I don't think it should get new requests at that point.

**Comment by fturib** [comment]
I am not very into the socket level.
But what I can see about the restart mechanisme is:
- 1- we duplicate the File Descriptiors (and below the sockets) of the "old" servers and connect the dup to the "new " servers to start. 
=> IMO, there is nothing that "close" the original sockets, so I would imagine that the "old" servers will continue to process msgs ...
- 2- we start the servers on the "new" sockets/FD => there are processing the incoming msgs.
- 3- we stop the "old" servers

QUESTION: when the msg are no longer arriving to the the "old" servers ? In other words, what happen to the incoming msg when you have duplicated sockets ?
- the last one win ? or the first one ?
- both sockets are processing the msg ? (in that case the msg would be processed twice)

I would say that both sockets would process the msg. Which make that the "old" servers will continue to serve the msg until it is really stopped. that means, whatever the graceful mechanism, the issue having the "OnShutdwn" event BEFORE the stop is a concern for plugins that have resources that needs to be alive until last msg is processed.

**Any thought on the way duplication of socket is working ?**

**Comment by rdrozhdzh** [comment]
I just tested duplicated UDP connections with simple application like https://play.golang.org/p/NMoLQreZ890 - this code is very similar to what we do  in coredns. My observations are
1. when I send individual requests (dig @127.0.0.1 -p 1053 test.com) I see 3 messages from Reader # 1 on each dig request (dig retries twice since it doesn't get response).
2. when I send many parallel queries with dnsperf, I see that some messages come to Reader # 1, and other messages go to Reader # 2
So, my conclusion - both servers can read messages from duplicated sockets, but any particular message is read only once - either by the first (original) server or by the second (duplicated) server

**Comment by johnbelamaric** [comment]
So we need to stop the server, then let it finish any outstanding queries. then destroy it.

**Comment by fturib** [comment]
Thanks for the test !

Now we need to design a solution:
1- either ensure an Event "Shutdown After Stop"
2- either each plugin deal with the OnShutdown event. For Foward/Procy it means do not accept to serve after the ressurce are deallocated in the OnShutdown.

What is weir is that the events raised by Caddy are not symetrical: of OnStart is BEFORE Start(), then OnShutdown should be AFTER Stop().

Best would be to have events Before and After Start/Stop. Depending the plugin, it may be more suitable to manage resource before of after the servers are started ...

**Comment by fturib** [comment]
@johnbelamaric : Do we want to recover after a failing Start of the new set of servers ?

**Comment by fturib** [comment]
NOTE: Anyway, right now, I think that if there is some Error raised during the Start process .. the final status is very uncertain. Some servers may be started (and share the socket), some other not.

We may better completely stopped the service if something wrong happen during starting and or stopping of the servers.

=> Which is different from validating the Corefile. As it is done BEFORE starting any new servers.

**Comment by rdrozhdzh** [comment]
Seems, if we want the _true solution_ it should be implemented on caddy's side.

As a workaround on coredns, we can develop a plugin which will have the highest priority, i.e. it will be the first handler in the chain. This plugin will count the queries in-progress, i.e. increment counter on enter to ServeDNS() and decrement on leaving. It will have an onShutdown handler which is expected to be called earlier than other onShutdown handlers. In this handler we will switch the chain to reject mode, i.e. ServeDNS() will return error and will not pass the messages to next plugin. After this, the handler will block until all in-progress queries leave the plugin chain. This way we will guarantee that no queries are in-progress at the time when other plugins run their onShutdown handlers

**Comment by miekg** [comment]
/cc @mholt
Matt: can we change the order is proposed in the initial comment?

**Comment by miekg** [comment]
alternatively; we can (I think) write our own reloader

**Comment by mholt** [comment]
@miekg Well, the logic of the current ordering is that if any of the shutdown callbacks return an error, then the closing of the listeners is aborted. If we reverse the logic then we have to be OK with the fact that shutdown callbacks can't abort if there's an error: they can only log it and the shutdown must go on.

Here is the relevant code: https://github.com/mholt/caddy/blob/88edca65d3b1b7fb533f5abe7c1a95a417a5d016/caddy.go#L231-L240

But maybe that is fine since the new listeners have started successfully anyway.

I'd accept a PR!

**Comment by fturib** [comment]
@rdrozhdzh : you wrote on a separate email thread : 
> we may need to improve coredns/caddy restart procedure to wait for purging messages from plugin chain before shutting them down

Do you think that having the event "onShutdown()" for each Plugin happen BEFORE the global Stop() of server will be enough ?

In other words : 
- can each plugin manage to purge the message by itself (and cleanup its resources) ? 
- or it needs something specific on the Server() to wait all msg are processed after OnShutdown() and before Stop() ?

**Comment by rdrozhdzh** [comment]
> can each plugin manage to purge the message by itself (and cleanup its resources) ?

This is what I did in my fix for forward plugin. But this is not ideal solution, because it drops the queries which come after plugin shutdown. Ideally, graceful server should not drop queries during reload.
Plugin can do something only to those queries which already reached the plugin handler. And plugin is not aware about those queries which were already pushed by server into beginning of plugin handlers chain but didn't reached the particular plugin yet.

> or it needs something specific on the Server() to wait all msg are processed after OnShutdown() and before Stop() ?

Normally queries should not be processed after at least one OnShutdown handler was called. Because some plugin might release some resources needed for query processing. 
Simple reordering of `plugins.onShutdown()` and `server.stop()` (pseudocode) will also not work. Because stopping the server (as it currently implemented) will likely close UDP connections and we won't be able to send response, i.e. we'll drop queries. Ideally, we should have the sequence like following
```
 server.stopReadingIncomingQueries() // let other instance to read
 server.waitForAllQueriesProcessed() // purge plugins chain
 plugins.onShutdown()
 server.stop()
```

**Comment by fturib** [comment]
So, if we split the work between Caddy code and CoreDNS Code:

*** ON Caddy side
- ensure the OnShutdoen() happen AFTER the servers are Stopped - so it is the symetric of "onStartuP" and we can deallocate safely the resources of the plugins

*** On CoreDNS:
- when Stop() a server : proceed by step:
1- close the incoming stream of queries
2- wait all queries are processed
3- stop the server (listener)

Would that work ?

**Comment by fturib** [comment]
Looking into the code of Server.go, the Stop() method is waiting for dnsWg. But that WaitingGroup is used nowhere ! (except the initial setup and the wait at end).

=> So I guess there is some missing code, or some code that has been removed.

```
func (s *Server) Stop() (err error) {
	if runtime.GOOS != "windows" {
		// force connections to close after timeout
		done := make(chan struct{})
		go func() {
			s.dnsWg.Done() // decrement our initial increment used as a barrier
			s.dnsWg.Wait()
			close(done)
		}()

		// Wait for remaining connections to finish or
		// force them all to close after timeout
		select {
		case <-time.After(s.connTimeout):
		case <-done:
		}
	}
```

And what make sense here, is to wait that all queries are processed (or timeout).
So, I am wondering if the missing code would be:


```
func (s *Server) Serve(l net.Listener) error {
	s.m.Lock()
	s.server[tcp] = &dns.Server{Listener: l, Net: "tcp", Handler: dns.HandlerFunc(func(w dns.ResponseWriter, r *dns.Msg) {
		s.dnsWg.Add(1)  <============== HERE
		defer s.dnsWg.Done()  <============== AND HERE
		ctx := context.WithValue(context.Background(), Key{}, s)
		s.ServeDNS(ctx, w, r)
	})}
	s.m.Unlock()

	return s.server[tcp].ActivateAndServe()
}

```

Doing so we are resolving our issue of stopping the server w/o waiting queries are processed.

@rdrozhdzh : can you confirm that would work ?

@miekg : do you remind the purpose of this dnsWg and if there is some code committed or removed that explain current situation ?

**Is there a drawback or some race condition issue doing so ?**

**Comment by fturib** [comment]
Hum.. that will not be enough : we need to inform the UDP and TCP listeners to not accept anymore the incoming request, but let the connection open in order to send the reply for in-process queries.

for UDP : We may be able to achieve that target using the `DecorateReader` that you can pass on the new Server. (or decorate the UDPConnection ...)
for TCP : we need to avoid to Accept() ...  => we may be able to decorate the TCP Listener in order to stop to Accept() while the server is shutting down.

**Comment by fturib** [comment]
hum...
is there really a way to NOT accept incoming request while finishing to process incoming ?
How can we implement the action ? 
```
server.stopReadingIncomingQueries() // let other instance to read
```

- That may be work for TCP =>  just do not Accept()
- But I do dot see for UDP : as long as the socket is connected, any packet can come in .. and if we disconnect, then we cannot write response of processed queries. So I do not see a way to finish processing on going queries while refusing the incoming ones.

**Comment by rdrozhdzh** [comment]
> @rdrozhdzh : can you confirm that would work ?

not exactly. If we call `Stop()` after the incoming message was already read but `Handler()` was not called yet (i.e. the query was not counted yet ) then server will stop and lose this query.

> for UDP : We may be able to achieve that target using the DecorateReader that you can pass on the new Server. (or decorate the UDPConnection ...)

hm, this may work. We can consider similar approach for established TCP connections with `DecorateReader.ReadTCP()`. 

For UDP we can just stop reading new messages from UDP (new server instance will read them). e.g. let `DecorateReader.ReadUDP()` return error if stop was initiated.

For TCP we should first do `TCPConn.CloseRead()` and then read all queued messages from the connection until EOF got

As for TCP listener, I think we can just close it at begin of `Stop()` and let new server instance accept new TCP connections

**Comment by fturib** [comment]
@rdrozhdzh : do you agree that this issue can be closed now that we merged the gracefull term from Miek/DNS servers ?

**Comment by rdrozhdzh** [comment]
I looked at the current code of `Restart()` in caddy and it still has inappropriate order of calling `instance.Stop()` and `onShutdown()` handlers. `Stop()` should be called before any `onShutdown()` to make the shutdown really graceful

https://github.com/mholt/caddy/blob/f7757da7edceb23de39e2460fd59068eb5834c93/caddy.go#L232-L242

Also, `Restart()` should call **all** `onShutdown()` handlers and return success (because we already have new instance running), even if some errors were returned by `Stop()` or some of `onShutdown()`. The errors can be logged instead.

**Comment by fturib** [comment]
@rdrozhdzh : you are right. I forgot the part tied to the order of the event in Caddy.
In this [comment](https://github.com/coredns/coredns/issues/1666#issuecomment-380624422), @mholt said he would accept a PR to re-order those events.

@rdrozhdzh : would you have time to build the PR for Caddy ?

**Comment by fturib** [comment]
I see that Caddy made a new release 0.11.1 3 dys ago.
I guess if we use that version instead of current 0.10.xx we can resolve this ticket.

WARNING: I  think we need thorough tests, the change of order of the events may have an impact ...

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>I see that Caddy made a new release 0.11.1 3 dys ago.
>I guess if we use that version instead of current 0.10.xx we can resolve this ticket.
>
>WARNING: I  think we need thorough tests, the change of order of the events may have an impact ...

submit PR - not too worried about this change.

**Comment by fturib** [comment]
@rdrozhdzh : now that PR #2321 is merged, upgrading Caddy to 0.11.1, I guess we are good with this issue, no ?

**Comment by rdrozhdzh** [comment]
yes, I think we can close the issue now
